### PR TITLE
Fix @Inject callback methods overriding a callback in a superclass

### PR DESCRIPTION
**Pull request by simon816**
### The (obscure) issue ###
A mixin to a class contains an injection, using `@Inject`. The injection is injecting into a constructor and so it has been named `onConstructed`.
E.g.
```java
public class Base {
    public Base() {
    }
}
```
```java
@Mixin(Base.class)
public class MixinBase {
    @Inject(method = "<init>", at = @At("RETURN"))
    public void onConstructed() {
        System.out.println("Base constructed");
    }
}
```
Another mixin is injecting into a different class, but the class has the previous's mixin target in it's hierarchy. This mixin also wants to inject into the constructor, so it defines an injection `onConstructed`.
E.g.
```java
public class Another extends Base {
}
```
```java
@Mixin(Another.class)
public class MixinAnother extends Base {
    @Inject(method = "<init>", at = @At("RETURN"))
    public void onConstructed() {
        System.out.println("Another constructed");
    }
}
```
Now that there are two identical methods, when creating a new instance the `onConstructed` method is only called in the child mixin.
E.g.
`new Another();`
STDOUT = `Another constructed`

The injection to the super class is never called

### Solution ###
Make the injected method private, always. Specifically, the `INVOKESPECIAL` will always reference the correct class (from my understanding) and so it won't override.

I do not know the exact place where it should be made private, so I just did it right before the invocation instruction is appended. Is it possible to change the access back to what it was (likely public) after appending the instruction?

### A real example in Sponge ###
Spawn a minecart, now call `minecart.getType()`. It will return `null` because the type is set in the `onConstructed` method in `MixinEntity`, but was overridden by `onConstructed` in `MixinEntityMinecart`.
(In fact, if you have the entity activation module active, it will error automatically because it calls `getType()` during a tick cycle. (Which is how I initially discovered the issue))

### CODE REVIEW

**Comment by Mumfrey** [comment]
This is supposed to already be handled, and I'm not really sure I like your solution.

It's already handled in that injected handlers are **supposed** to be allowed to be overridden in derived mixins, this is why the injector injects `INVOKESPECIAL` if the handler method is **private** and `INVOKEVIRTUAL` if the handler method is public, protected, or package-private. The intention is that it's left up to the mixin author to decide whether derived mixins should be allowed to override an injected handler.

For example, in the instance here, the actual recommended solutions are:

* Have the derived mixin override the method and call `super` as appropriate and remove the explicit injection in the derived class
* Have the super mixin declare the handler as private, this will inject an `INVOKESPECIAL` instead and call the private method, forcing the derived mixin to have its own handler

Your other assertion:

> Make the injected method private, always. Specifically, the INVOKESPECIAL will always reference the correct class (from my understanding) and so it won't override.

Is not really correct. In java 6, this is partly correct, and is the reason for all my static binding resolution code, but in java 7 a best-effort search (which basically does what mine does, albeit in the JVM rather than in code) is made for the "right" method to reference given the hierarchy. The semantics would work as you'd expect if the method is `private`, but as I mentioned, this is already the case!

Basically, this works as intended. Injected handlers can be overridden where specified (by simply making them protected, package-private, or public) or not (by making them private) and the appropriate opcode is injected in the target method based on the visibility of the handler method. Removing this ability simply means it's not possible to design in (to a mixin) the ability to have over-ridable injections and thus force derived classes to make their own (redundant) injections.

**Comment by simon816** [comment]
OK. I was not aware this was intended functionality. It would be difficult when writing a mixin to know whether the injection you've just wrote has overridden another, or if another has just overridden yours.
Because mixins don't have to extend each other, enforcing `@Override` is not an option, however is it possible that there could be some safety about injections overriding each other? Perhaps an error could be raised if an injection is overriding without some sort of `@OverrideInjection` annotation

**Comment by Mumfrey** [comment]
Yes potentially, or another use for `@SoftOverride` perhaps.
